### PR TITLE
Restore graceful shutdown

### PR DESCRIPTION
**Pull request by JekaMas**

### CODE REVIEW

File: eth/downloader/downloader.go [hunk 1]
```diff
@@ -1300,6 +1306,12 @@ func (d *Downloader) fetchParts(deliveryCh chan dataPack, deliver func(dataPack)
 			}
 			// Check for fetch request timeouts and demote the responsible peers
 			for pid, fails := range expire() {
+				select {
```
Comment (AlexeyAkhunov) [inline eth/downloader/downloader.go:1309]: Is this needed? The loop is not doing any heavy lifting, and will probably execute almost instantly

File: eth/downloader/downloader.go [hunk 2]
```diff
@@ -1241,6 +1245,8 @@ func (d *Downloader) fetchParts(deliveryCh chan dataPack, deliver func(dataPack)
 	finished := false
 	for {
 		select {
+		case <-d.quitCh:
```
Comment (AlexeyAkhunov) [inline eth/downloader/downloader.go:1248]: Termination always cancels too, so this is not required

File: eth/downloader/downloader.go [hunk 3]
```diff
@@ -1347,6 +1359,12 @@ func (d *Downloader) fetchParts(deliveryCh chan dataPack, deliver func(dataPack)
 			idles, total := idle()
 
 			for _, peer := range idles {
+				select {
```
Comment (AlexeyAkhunov) [inline eth/downloader/downloader.go:1362]: This is inconsistent. Why not to just check cancelCh, or perhaps just not to anything here and rely on fetch function and the outer loop to do the job

File: eth/downloader/stagedsync_downloader.go [hunk 4]
```diff
@@ -123,3 +105,67 @@ func (d *Downloader) doStagedSyncWithFetchers(p *peerConnection, headersFetchers
...
+	}
+	log.Info("Checking for unwinding... Complete!")
+	return nil
+}
+
+func (d *Downloader) runStep(fn func() error) error {
+	return run(d.quitCh, fn)
+}
+
+func run(quitCh chan struct{}, fn func() error) error {
```
Comment (AlexeyAkhunov) [inline eth/downloader/stagedsync_downloader.go:164]: Does this function really add value for extra code complexity it introduces? It only checks the channel once at the beginning anyway, which is not really the problem. We are concerned with the interrupting in the middle of long actions

File: eth/downloader/stagedsync_downloader.go [hunk 5]
```diff
@@ -123,3 +105,67 @@ func (d *Downloader) doStagedSyncWithFetchers(p *peerConnection, headersFetchers
...
+
+		if err = d.runStep(stageFn); err != nil {
+			return fmt.Errorf("error unwinding stage: %d: %v", stage, err)
+		}
+	}
+	log.Info("Checking for unwinding... Complete!")
+	return nil
+}
+
+func (d *Downloader) runStep(fn func() error) error {
```
Comment (AlexeyAkhunov) [inline eth/downloader/stagedsync_downloader.go:160]: Don't think this add value, please remove

File: eth/downloader/downloader.go [hunk 6]
```diff
@@ -1520,7 +1538,7 @@ func (d *Downloader) processHeaders(origin uint64, pivot uint64, td *big.Int) er
 						n, newCanonical, lowestCanonicalNumber, err = d.blockchain.InsertHeaderChainStaged(chunk, frequency)
 						if newCanonical {
 							// Need to unwind further stages
-							if err1 := UnwindAllStages(d.stateDB, lowestCanonicalNumber); err1 != nil {
+							if err1 := UnwindAllStages(d.stateDB, lowestCanonicalNumber, d.quitCh); err1 != nil {
```
Comment (AlexeyAkhunov) [inline eth/downloader/downloader.go:1541]: This is not a long running function, and it needs to execute atomically, giving it interruptibility is overkill and is not required

File: common/etl/etl.go [hunk 7]
```diff
@@ -269,8 +275,22 @@ func readElementFromDisk(decoder Decoder) ([]byte, []byte, error) {
...
 	bucket []byte
+	quit chan struct{}
 }
 
 func (s *bucketState) Get(key []byte) ([]byte, error) {
 	return s.getter.Get(s.bucket, key)
 }
+
+
+func (s *bucketState) Stopped() error {
```
Comment (mandrigin) [inline common/etl/etl.go:286]: I would just use `bool` as a return value there, but otherwise looks good.

Comment (mandrigin) [reply common/etl/etl.go:286]: or, maybe we can have a global ErrStopped so we handle that same shutdown the same way?

Comment (JekaMas) [reply common/etl/etl.go:286]: I think we can extract this into a small common packsge

**Pull request merged**